aten/src/ATen/native/sparse/SparseCsrTensorMath.cpp

```diff
@@ -1128,5 +1130,136 @@ Tensor _sparse_csr_prod_cpu(const Tensor& input, IntArrayRef dims_to_reduce, boo
+      tiled_sizes.push_back(blocksize[0]);
+      tiled_sizes.push_back(size_neg_1_blocked);
+      tiled_sizes.push_back(blocksize[1]);
+      return t.reshape(tiled_sizes).transpose(-3, -2);
+    }
+  };
+
+  // Note that sparse values are (..., b0, b1). This means that
+  // the strided input has to be "tilable" to (..., b1, x) with
+  // any x >= 1 such that all the shapes are (block) matrix product
+  // compatible. The matrix product will then have shape (..., b0, x).
+  // This in turn means the the result has to be "tilable" to
+  // (..., b0, x).
+  //
+  // These observations imply the following restrictions:
+  // 1. strided.size(-2) has to be divisible by b1.
+  // 2. result.size(-2) has to be divisible by b0.
+  // 3. both strided.size(-1) and result.size(-1)
+  //    have to be divisible by x.
+  //
+  // Restriction 2 is trivially satisfied.
+  // Regarding restriction 3:
+  // it would make sense to take the largest possible x for better
+  // performance since it is very likely that the last dimension
+  // is contiguous. As such, this value is exactly
+  // x = strided.size(-1), since strided.size(-1) == result.size(-1)
+
+  // See the comments above. This is our x.
+  const auto outer_blocksize = n;
+
+  Blocksize strided_blocksize = {blocksize[1], outer_blocksize};
+  const auto strided_tiled = tile_tensor(strided, strided_blocksize);
+
+  // Left argument is (..., b0, b1) and right is (..., b1, x).
+  // This naturally implies the result should be "tilable" as
+  // (..., b0, x).
+  Blocksize result_blocksize = {blocksize[0], outer_blocksize};
+  auto result_tiled = tile_tensor(result, result_blocksize);
+
+  if (compressed_layout == kSparseCsr) {
+    values.unsqueeze_(-1).unsqueeze_(-1);
+  }
+
+  Tensor compressed_indices, plain_indices;
+  std::tie(compressed_indices, plain_indices) = at::sparse_csr::getCompressedPlainIndices(compressed);
+
+  // Select blocked rows of the strided input that intersect with the blocked colums of the sparse input.
+  auto strided_tiled_selected_rows = strided_tiled.index_select(-4, plain_indices);
+
+  // Now that we know which block rows intersect with which block columns,
+  // we can perform matrix produts between pairs of blocks.
+  const auto pairwise_block_mm = values.unsqueeze(-3).matmul(strided_tiled_selected_rows);
+
+  // Having pairwise block matrix products stored in pairwise_block_mm,
+  // it is sufficient to sum all the block products that share the same row
+  // incoded in the sparse index. Since the reduction step is done via
+  // advanced indexing methods, the compressed index ought to get converted
+  // to the COO format.
+  const auto compressed_indices_coo = at::_convert_indices_from_csr_to_coo(
+      compressed_indices,
+      plain_indices,
+      compressed_indices.scalar_type() == kInt).select(0, 0);
+
+  // Reduction step.
+  // Zero out and sum over the blocks that share the same row indices.
+  result_tiled.zero_();
```

Comment:
Note: Once handling of alpha/beta are added you will need to account for `beta==zero` as a special case. See [addmm docs](https://pytorch.org/docs/stable/generated/torch.addmm.html?highlight=addmm#torch.addmm)  for the details!